File: app/src/main/java/com/hwonchul/movie/util/BindingUtils.java
package com.hwonchul.movie.util;

import android.graphics.drawable.Drawable;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.BindingAdapter;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager2.widget.ViewPager2;

import com.bumptech.glide.Glide;
import com.bumptech.glide.signature.ObjectKey;
import com.hwonchul.movie.R;
import com.hwonchul.movie.domain.model.Movie;
import com.hwonchul.movie.domain.model.Image;
import com.hwonchul.movie.domain.model.Video;
import com.hwonchul.movie.presentation.view.details.PosterThumbnailAdapter;
import com.hwonchul.movie.presentation.view.details.VideoAdapter;
import com.hwonchul.movie.presentation.view.listing.MovieAdapter;
import com.hwonchul.movie.presentation.view.poster.PosterAdapter;

import java.time.LocalDate;
import java.util.List;

import timber.log.Timber;

public class BindingUtils {

    @BindingAdapter(value = {"imageUrl", "placeholder", "fallback"}, requireAll = false)
    public static void bindImage(@NonNull ImageView view,
                                 @NonNull String url,
                                 @Nullable Drawable placeholder,
                                 @Nullable Drawable fallback) {
        Glide.with(view.getContext())
                .load(url)
                // key를 하루 단위로 변경한다.
                .signature(new ObjectKey(LocalDate.now().toEpochDay()))
                .placeholder(placeholder)
                .fallback(fallback)
                .into(view);
    }

    @BindingAdapter(value = {"listItem", "onClick"})
    public static void bindMovieListItems(@NonNull RecyclerView listView,
                                          @Nullable List<Movie> items,
                                          @NonNull MovieAdapter.OnMovieDetailListener listener) {
        MovieAdapter adapter = (MovieAdapter) listView.getAdapter();
        if (adapter == null) {
            adapter = new MovieAdapter();
            // layout Manager
            listView.setLayoutManager(new LinearLayoutManager(
                    listView.getContext(), LinearLayoutManager.VERTICAL, false));
            // 구분선 세팅
            final int outSideSpace = 0;
            final int divHeight = listView.getContext()
                    .getResources()
                    .getDimensionPixelSize(R.dimen.divider_large_16dp);
            listView.addItemDecoration(new VerticalRecyclerViewDecoration(outSideSpace, divHeight));
            // listener
            adapter.setOnMovieDetailListener(listener);
            listView.setAdapter(adapter);
        }
        // item 세팅
        if (items != null) {
            adapter.setItems(items);
        }
    }

    @BindingAdapter(value = {"status"})
    public static void bindStatus(@NonNull ImageView view,
                                          @Nullable String status) {
        if(status == null) {
            return;
        }

        int image;
        switch (status) {
            case "Released":
                image = R.drawable.ic_released;
                break;
            default:
                image = R.drawable.ic_production;
                break;
        }
        view.setImageResource(image);
    }

    @BindingAdapter(value = {"listItem", "listener"})
    public static void bindPosterThumbnailList(@NonNull RecyclerView listView,
                                               @Nullable List<Image> items,
                                               @NonNull PosterThumbnailAdapter.OnClickListener listener) {
        PosterThumbnailAdapter adapter = (PosterThumbnailAdapter) listView.getAdapter();
        if (adapter == null) {
            adapter = new PosterThumbnailAdapter();

            // layout Manager
            listView.setLayoutManager(new LinearLayoutManager(
                    listView.getContext(), LinearLayoutManager.HORIZONTAL, false));
            // 구분선 세팅
            final int outSideSpace = listView.getContext()
                    .getResources()
                    .getDimensionPixelSize(R.dimen.divider_large_16dp);
            final int divWidth = listView.getContext()
                    .getResources()
                    .getDimensionPixelSize(R.dimen.divider_small_4dp);
            listView.addItemDecoration(new HorizontalRecyclerViewDecoration(outSideSpace, divWidth));
            // listener
            adapter.setOnClickListener(listener);
            listView.setAdapter(adapter);
        }
        // items 세팅
        if (items != null) {
            adapter.setItems(items);
        }
    }

    @BindingAdapter(value = {"listItem", "listener"})
    public static void bindVideoList(@NonNull RecyclerView listView,
                                     @Nullable List<Video> items,
                                     @NonNull VideoAdapter.OnClickListener listener) {
        VideoAdapter adapter = (VideoAdapter) listView.getAdapter();
        if (adapter == null) {
            adapter = new VideoAdapter();

            // layout Manager
            listView.setLayoutManager(new LinearLayoutManager(
                    listView.getContext(), LinearLayoutManager.HORIZONTAL, false));
            // 구분선 세팅
            final int outSideSpace = listView.getContext()
                    .getResources()
                    .getDimensionPixelSize(R.dimen.divider_large_16dp);
            final int divWidth = listView.getContext()
                    .getResources()
                    .getDimensionPixelSize(R.dimen.divider_small_4dp);
            listView.addItemDecoration(new HorizontalRecyclerViewDecoration(outSideSpace, divWidth));
            // listener
            adapter.setOnClickListener(listener);
            listView.setAdapter(adapter);
        }
        // items 세팅
        if (items != null) {
            adapter.setItems(items);
        }
    }

    @BindingAdapter(value = {"listItem", "currentPos"})
    public static void bindPosterList(@NonNull ViewPager2 viewpager,
                                      @Nullable List<Image> items,
                                      @Nullable Integer currentPos) {
        PosterAdapter adapter = (PosterAdapter) viewpager.getAdapter();
        if (adapter == null) {
            adapter = new PosterAdapter();
        }
        // item 세팅
        if (items != null) {
            adapter.setItems(items);
            Timber.d("PosterAdapter Items : %s", items.size());
        }
        viewpager.setAdapter(adapter);
        viewpager.setOffscreenPageLimit(5);
        if (currentPos != null) {
            viewpager.setCurrentItem(currentPos, false);
            Timber.d("PosterAdapter Current Pos : %s", currentPos);
        }
    }
}

File: app/src/main/java/com/hwonchul/movie/util/BindingUtils.kt
package com.hwonchul.movie.util

import android.graphics.drawable.Drawable
import android.widget.ImageView
import androidx.databinding.BindingAdapter
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.widget.ViewPager2
import com.bumptech.glide.Glide
import com.bumptech.glide.signature.ObjectKey
import com.hwonchul.movie.R
import com.hwonchul.movie.domain.model.Image
import com.hwonchul.movie.domain.model.Movie
import com.hwonchul.movie.domain.model.Video
import com.hwonchul.movie.presentation.details.PosterThumbnailAdapter
import com.hwonchul.movie.presentation.details.VideoAdapter
import com.hwonchul.movie.presentation.details.poster.PosterAdapter
import com.hwonchul.movie.presentation.home.MovieAdapter
import com.hwonchul.movie.presentation.home.MovieAdapter.OnMovieDetailListener
import timber.log.Timber
import java.time.LocalDate

object BindingUtils {

    @JvmStatic
    @BindingAdapter(value = ["imageUrl", "placeholder", "fallback"], requireAll = false)
    fun bindImage(
        view: ImageView,
        url: String?,
        placeholder: Drawable?,
        fallback: Drawable?
    ) {
        Glide.with(view.context)
            .load(url)
            // key를 하루 단위로 변경한다.
            .signature(ObjectKey(LocalDate.now().toEpochDay()))
            .placeholder(placeholder)
            .fallback(fallback)
            .into(view)
    }

    @JvmStatic
    @BindingAdapter(value = ["listItem", "onClick"])
    fun bindMovieListItems(
        listView: RecyclerView,
        items: List<Movie>?,
        listener: OnMovieDetailListener
    ) {
        (listView.adapter as MovieAdapter? ?: MovieAdapter().also { adapter ->
            // layout Manager
            listView.layoutManager = LinearLayoutManager(
                listView.context, LinearLayoutManager.HORIZONTAL, false
            )

            // 구분선 세팅
            val outSideSpace = listView.context
                .resources
                .getDimensionPixelSize(R.dimen.divider_large_16dp)
            val divHeight = listView.context
                .resources
                .getDimensionPixelSize(R.dimen.divider_large_16dp)
            listView.addItemDecoration(HorizontalRecyclerViewDecoration(outSideSpace, divHeight))

            // listener
            adapter.setOnMovieDetailListener(listener)
            listView.adapter = adapter
        }).apply { items?.let { setItems(it) } }
    }

    @JvmStatic
    @BindingAdapter(value = ["status"])
    fun bindStatus(
        view: ImageView,
        status: String?
    ) {
        if (status == null) {
            return
        }
        val image = when (status) {
            "Released" -> R.drawable.ic_released
            else -> R.drawable.ic_production
        }
        view.setImageResource(image)
    }

    @JvmStatic
    @BindingAdapter(value = ["listItem", "listener"])
    fun bindPosterThumbnailList(
        listView: RecyclerView,
        items: List<Image>?,
        listener: PosterThumbnailAdapter.OnClickListener
    ) {
        (listView.adapter as PosterThumbnailAdapter? ?: PosterThumbnailAdapter().also { adapter ->
            // layout Manager
            listView.layoutManager = LinearLayoutManager(
                listView.context, LinearLayoutManager.HORIZONTAL, false
            )

            // 구분선 세팅
            val outSideSpace = listView.context
                .resources
                .getDimensionPixelSize(R.dimen.divider_large_16dp)
            val divWidth = listView.context
                .resources
                .getDimensionPixelSize(R.dimen.divider_small_4dp)
            listView.addItemDecoration(HorizontalRecyclerViewDecoration(outSideSpace, divWidth))

            // listener
            adapter.setOnClickListener(listener)
            listView.adapter = adapter
        }).apply { items?.let { setItems(it) } }
    }

    @JvmStatic
    @BindingAdapter(value = ["listItem", "listener"])
    fun bindVideoList(
        listView: RecyclerView,
        items: List<Video>?,
        listener: VideoAdapter.OnClickListener
    ) {
        (listView.adapter as VideoAdapter? ?: VideoAdapter().also { adapter ->
            // layout Manager
            listView.layoutManager = LinearLayoutManager(
                listView.context, LinearLayoutManager.HORIZONTAL, false
            )

            // 구분선 세팅
            val outSideSpace = listView.context
                .resources
                .getDimensionPixelSize(R.dimen.divider_large_16dp)
            val divWidth = listView.context
                .resources
                .getDimensionPixelSize(R.dimen.divider_small_4dp)
            listView.addItemDecoration(HorizontalRecyclerViewDecoration(outSideSpace, divWidth))

            // listener
            adapter.setOnClickListener(listener)
            listView.adapter = adapter
        }).apply { items?.let { setItems(it) } }
    }

    @JvmStatic
    @BindingAdapter(value = ["listItem", "currentPos"])
    fun bindPosterList(
        viewpager: ViewPager2,
        items: List<Image>?,
        currentPos: Int?
    ) {
        (viewpager.adapter as PosterAdapter? ?: PosterAdapter().also { adapter ->
            viewpager.adapter = adapter
            viewpager.offscreenPageLimit = 5
        }).apply {
            items?.let {
                setItems(it)
                Timber.d("PosterAdapter Items : %s", it.size)
            }
            currentPos?.let {
                viewpager.setCurrentItem(it, false)
                Timber.d("PosterAdapter Current Pos : %s", it)
            }
        }
    }
}